### PR TITLE
[Profiler][PrivateUse1] Profiler support PrivateUse1 key

### PR DESCRIPTION
**Pull request by fwenguang**
Summary:
1.Package public headers of kineto if USE_KINETO so that they can be used by PrivateUse1 user.
2.Add PrivateUse1 key to ActivityType.
3. Support PrivateUse1 key in function deviceTypeFromActivity and _supported_activities.
4. Fix some bugs when processing profiler results.

### CODE REVIEW

**Comment by fwenguang** [comment]
Depend on https://github.com/pytorch/kineto/pull/874

File: torch/autograd/profiler.py [hunk 1]
```diff
@@ -292,6 +292,14 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
         if self.use_cuda:
             torch.cuda.synchronize()
+        if self.use_device:
+            try:
+                privateuse1_module = getattr(torch, self.use_device)
+                privateuse1_module.synchronize()
+            except AttributeError:
+                warn(
+                    f"Not found torch.{self.use_device}.synchronize(). Skip syncing on {self.use_device}."
+                )
```
Comment (albanD) [inline torch/autograd/profiler.py:302]:
```suggestion
            privateuse1_module = getattr(torch, self.use_device)
            if hasattr(privateuse1_module, "synchronize"):
                privateuse1_module.synchronize()
```

**Review by albanD — COMMENTED** [review]
nit to avoid the extra exception.

**Comment by aaronenyeshi** [comment]
Adding @xunsongh, to help with cross-review, since privateuse1 and xpu are going through similar code.

**Comment by xunsongh** [comment]
Thank you for bringing this to my attention. Based on my review, it appears that the modifications in this PR do not affect the XPU pathway, so I have no objections. Thank you.

**Review by aaronenyeshi — CHANGES_REQUESTED** [review]
Please take a look at failing CI

**Comment by fwenguang** [comment]
> Please take a look at failing CI

This patch needs `libkineto::ActivityType::PRIVATEUSE1_RUNTIME`.  The pr https://github.com/pytorch/kineto/pull/874 in kineto should be merged first.

**Comment by fwenguang** [comment]
Hi @aaronenyeshi , thank you for accepting the dependented kineto pr https://github.com/pytorch/kineto/pull/874.  I just updated commit id of kineto, please also help to review this.

File: torch/autograd/profiler_util.py [hunk 2]
```diff
@@ -271,6 +271,12 @@ def supported_export_stacks_metrics(self):
         ]
 
     def export_stacks(self, path: str, metric: str):
+        # Support passing more meaningful names for privateuse1 users
+        metric = (
+            metric.replace(self._use_device, "privateuse1")
```
Comment (aaronenyeshi) [inline torch/autograd/profiler_util.py:276]: should _use_device be more generic, since we may have other devices in the future?

**Review by aaronenyeshi — COMMENTED** [review]
Is it possible to make use_device a generic string, so that various devices can use it?

File: torch/autograd/profiler_util.py (hunk 2 above)
Comment (fwenguang) [reply torch/autograd/profiler_util.py:276]: Without this replacing, metric only can be set to "self_privateuse1_time_total" for privateuse1 backend. but our user may pass a more meaningful name.
For example, our privateuse1 backend named "MLU", so self._use_device will be set to "mlu".  
We prefer to set metric to  "self_mlu_time_total", there will raise an error if no replacing. 
Of course, we can also use "self_privateuse1_time_total" , in which case replacing will do nothing.

For other devices, their self._user_device may be some ohter strings,  they can use the original string "self_privateuse1_time_total" as metric and also can use self_{self._use_device}_time_total. The logical is generic.

So, i believe self._use_device should be specific, like the current design, allowing it to represent different device names. Additionally, we can use self._use_device  to extend some more generic strings like self_privateuse1_time_total to specific.

Comment (fwenguang) [reply torch/autograd/profiler_util.py:277]: Hi, @aaronenyeshi @briancoutinho , could you please give some suggestions about this?

**Review by aaronenyeshi — COMMENTED** [review]
Hi @fwenguang , please check out https://github.com/pytorch/pytorch/pull/123247, which plans to unify the various devices for self.use_device. I think that would be a great way to get all of these devices on the same path for readability and maintainability.

**Comment by aaronenyeshi** [comment]
There is also a windows build failure: 
```
[6093/7932] Building CXX object third_party\kineto\libkineto\CMakeFiles\kineto_base.dir\src\output_json.cpp.obj
FAILED: third_party/kineto/libkineto/CMakeFiles/kineto_base.dir/src/output_json.cpp.obj 
sccache C:\PROGRA~2\MICROS~1\2019\BUILDT~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\cl.exe  /nologo /TP -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_EXTERNAL_MZCRC -D_CRT_SECURE_NO_DEPRECATE=1 -IC:\actions-runner\_work\pytorch\pytorch\cmake\..\third_party\benchmark\include -IC:\actions-runner\_work\pytorch\pytorch\third_party\onnx -IC:\actions-runner\_work\pytorch\pytorch\build\third_party\onnx -IC:\actions-runner\_work\pytorch\pytorch\third_party\foxi -IC:\actions-runner\_work\pytorch\pytorch\build\third_party\foxi -IC:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\include -IC:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\src -IC:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\third_party\dynolog -IC:\actions-runner\_work\pytorch\pytorch\third_party\fmt\include -IC:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\third_party\dynolog\dynolog\src\ipcfabric -I\extras\CUPTI\include -I\include -I\include\roctracer -external:I C:\actions-runner\_work\pytorch\pytorch\build\third_party\gloo -external:I C:\actions-runner\_work\pytorch\pytorch\cmake\..\third_party\gloo -external:I C:\actions-runner\_work\pytorch\pytorch\cmake\..\third_party\googletest\googlemock\include -external:I C:\actions-runner\_work\pytorch\pytorch\cmake\..\third_party\googletest\googletest\include -external:I C:\actions-runner\_work\pytorch\pytorch\third_party\protobuf\src -external:I C:\Jenkins\Miniconda3\Library\include -external:I C:\actions-runner\_work\pytorch\pytorch\third_party\XNNPACK\include -external:I C:\actions-runner\_work\pytorch\pytorch\third_party\ittapi\include -external:I C:\actions-runner\_work\pytorch\pytorch\cmake\..\third_party\eigen -external:I C:\actions-runner\_work\pytorch\pytorch\third_party\ideep\mkl-dnn\include\oneapi\dnnl -external:I C:\actions-runner\_work\pytorch\pytorch\third_party\ideep\include -external:W0 /DWIN32 /D_WINDOWS /GR /EHsc /Zc:__cplusplus /bigobj /FS /utf-8 -DUSE_PTHREADPOOL -DNDEBUG /O2 /Ob2 /DNDEBUG /bigobj -DNDEBUG -MD -DMKL_HAS_SBGEMM -DCAFFE2_USE_GLOO -DKINETO_NAMESPACE=libkineto -DFMT_HEADER_ONLY -DENABLE_IPC_FABRIC /std:c++17 -DWIN32_LEAN_AND_MEAN -DNOGDI -std:c++17 /showIncludes /Fothird_party\kineto\libkineto\CMakeFiles\kineto_base.dir\src\output_json.cpp.obj /Fdthird_party\kineto\libkineto\CMakeFiles\kineto_base.dir\ /FS -c C:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\src\output_json.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include\memory(3089): error C2027: use of undefined type 'libkineto::ActivityBuffers'
C:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\include\output_base.h(25): note: see declaration of 'libkineto::ActivityBuffers'
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include\memory(3088): note: while compiling class template member function 'void std::default_delete<libkineto::ActivityBuffers>::operator ()(_Ty *) noexcept const'
        with
        [
            _Ty=libkineto::ActivityBuffers
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include\memory(3198): note: see reference to function template instantiation 'void std::default_delete<libkineto::ActivityBuffers>::operator ()(_Ty *) noexcept const' being compiled
        with
        [
            _Ty=libkineto::ActivityBuffers
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include\memory(3125): note: see reference to class template instantiation 'std::default_delete<libkineto::ActivityBuffers>' being compiled
C:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\src\output_json.cpp(600): note: see reference to class template instantiation 'std::unique_ptr<libkineto::ActivityBuffers,std::default_delete<libkineto::ActivityBuffers>>' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include\memory(3089): error C2338: can't delete an incomplete type
C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include\memory(3090): warning C4150: deletion of pointer to incomplete type 'libkineto::ActivityBuffers'; no destructor called
C:\actions-runner\_work\pytorch\pytorch\third_party\kineto\libkineto\include\output_base.h(25): note: see declaration of 'libkineto::ActivityBuffers'
```

**Review by aaronenyeshi — COMMENTED** [review]
Overall looks good, let's wait for https://github.com/pytorch/pytorch/pull/123247 to land, and rebase on top of it, is that okay?

File: torch/autograd/profiler_util.py (hunk 2 above)
Comment (fwenguang) [reply torch/autograd/profiler_util.py:277]: Okay, i will revert this.


```python
                    metric_value = getattr(
                        evt,
                        metric.replace("cuda", "device")
                        .replace("xpu", "device")
                        .replace("privateuse1", "device"),
                    )
```
By the way, i noticed the above code in https://github.com/pytorch/pytorch/pull/123247/files,   Actually, this modification doesn't solve the issue I wanted to address. What I'm looking to do is something like appending a new line `.replace(self._use_device, "device")`. However, this isn't a critical issue, just something for convenience.  For now, I'll remove my modification.

**Comment by fwenguang** [comment]
> Overall looks good, let's wait for #123247 to land, and rebase on top of it, is that okay?

Okay，thank you.

**Review by aaronenyeshi — DISMISSED** [review]
LGTM!

**Comment by sraikund16** [comment]
This PR has caused a regression particularly on this line: https://github.com/pytorch/kineto/blob/main/libkineto/src/CuptiActivityProfiler.cpp#L1170

We are seeing traces that do not enter the conditional even though it should and it causes the indexes to be formatted incorrectly 

cc @aaronenyeshi

**Comment by aaronenyeshi** [comment]
@fwenguang , sorry about this. This diff is causing a regression to the chrome trace UI of GPU traces (GPU processes are no longer named or sorted after CPU processes). We will revert this, please take a look and re-open a new PR. 

@pytorchbot revert -m "Breaks GPU Chrome trace UI" -c ignoredsignal

Before:
image
After:
image

**Comment by fwenguang** [comment]
> @fwenguang , sorry about this. This diff is causing a regression to the chrome trace UI of GPU traces for PyTorch Profiler users (GPU processes are no longer named or sorted after CPU processes). We will revert this, please take a look and re-open a new PR.
> 
> Before: image After: image

Hi, @aaronenyeshi , I previously added `libkineto::ActivityType::PRIVATEUSE1_RUNTIME` into `kCpuTypes`, which was right, but it's no longer possible to detemine if it's PrivateUse1 backend through `libkineto::hasPrivateUse1Type `, i have removed `hasPrivateUse1Type`, see PR:https://github.com/pytorch/kineto/pull/909, once it is merged and updated, this PR can be merged without any modifications.

**Review by Youcantstopme2744 — CHANGES_REQUESTED** [review]
Need changes

**Comment by Youcantstopme2744** [comment]
> Summary:
> 1.Package public headers of kineto if USE_KINETO so that they can be used by PrivateUse1 user.
> 2.Add PrivateUse1 key to ActivityType.
> 3. Support PrivateUse1 key in function deviceTypeFromActivity and _supported_activities.
> 4. Fix some bugs when processing profiler results.

**Comment by aaronenyeshi** [comment]
Moving to new PR

**Pull request closed without merging**